### PR TITLE
bpo-35560: Remove incorrect assert statement that caused segfault in debug builds

### PR DESCRIPTION
**Pull request by tirkarthi**
* Removed assert statement checking for negative `min_width`. This caused segfault in debug builds for some format string in float. Reference for negative `min_width` : https://github.com/python/cpython/blob/59423e3ddd736387cef8f7632c71954c1859bed0/Python/formatter_unicode.c#L529
* Add tests from https://bugs.python.org/msg99839

https://bugs.python.org/issue35560

### CODE REVIEW

File: Lib/test/test_float.py
Comment (serhiy-storchaka) [inline Lib/test/test_float.py]: `float(123)` is the same as `123.0`. This line is not needed.

The following lines do not add anything for testing purpose too. `str.format()` and f-strings just use `format()` internally.

But it may be worth to add tests for format strings containing explicit format char '00f', '00e', '00g', and containing precision: '00.20', '0020f', etc. Test also with negative number, -123.0.

File: Misc/NEWS.d/next/Core and Builtins/2018-12-22-22-19-51.bpo-35560.9vMWSP.rst
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Core and Builtins/2018-12-22-22-19-51.bpo-35560.9vMWSP.rst]: This is too general and unlikely useful. Please specify place/conditions of the segfault.

Comment (tirkarthi) [reply Misc/NEWS.d/next/Core and Builtins/2018-12-22-22-19-51.bpo-35560.9vMWSP.rst]: Added a note on the function and assert statement. Thanks.

File: Lib/test/test_float.py
Comment (tirkarthi) [reply Lib/test/test_float.py]: * Added test cases as suggested.
* Added `min_width = Py_MAX(0, min_width);` as suggested in https://bugs.python.org/msg332359.

File: Lib/test/test_float.py
Comment (serhiy-storchaka) [inline Lib/test/test_float.py]: The bug is not reproducible with non-zero width. Add instead tests for non-zero precision and other format characters:  '00.10e', '00.10g'.

File: Lib/test/test_float.py
Comment (serhiy-storchaka) [inline Lib/test/test_float.py]: 123.340 is the same as 123.34.

File: Misc/NEWS.d/next/Core and Builtins/2018-12-22-22-19-51.bpo-35560.9vMWSP.rst
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Core and Builtins/2018-12-22-22-19-51.bpo-35560.9vMWSP.rst]: `min_width` and `_PyUnicode_InsertThousandsGrouping` are implementation details. Describe the issue in terms of Python.

File: Misc/NEWS.d/next/Core and Builtins/2018-12-22-22-19-51.bpo-35560.9vMWSP.rst
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Core and Builtins/2018-12-22-22-19-51.bpo-35560.9vMWSP.rst]: What are certain cases? This is a concrete case: floating point formatting with zero padding and small width.

Comment (tirkarthi) [reply Misc/NEWS.d/next/Core and Builtins/2018-12-22-22-19-51.bpo-35560.9vMWSP.rst]: Is the below content okay? It will be helpful if you have any in mind so that I can use it since I don't know how to explain it without the internals in the news entry.

```rst
Removed an incorrect assert statement that caused segfault in :func:`format`
in debug builds for floating point formatting with zero padding and small width. 
Patch by Karthikeyan Singaravelan.
```

Comment (serhiy-storchaka) [reply Misc/NEWS.d/next/Core and Builtins/2018-12-22-22-19-51.bpo-35560.9vMWSP.rst]: It LGTM, but native speakers can suggest better wording.

**Comment by tirkarthi** [comment]
@vstinner Friendly ping if you can review this or if you have a better fix in mind since you wrote the original version I will be happy to close the PR.

File: Objects/unicodeobject.c
Comment (vstinner) [inline Objects/unicodeobject.c:9384]: I would prefer to fix the caller of this function instead: Python/formatter_unicode.c.

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Objects/unicodeobject.c
Comment (tirkarthi) [reply Objects/unicodeobject.c:9384]: Thanks Victor. Few clarifications : 

* By fixing the caller do you mean `min_width` as in `min_width = Py_MAX(0, spec->n_min_width);` and passing it to `_PyUnicode_InsertThousandsGrouping` from Python/formatter_unicode.c without modifying spec->n_min_width ? There is an explicit comment that `spec->n_min_width` and I don't want to mutate the `spec->n_min_width` . 

* `_PyUnicode_InsertThousandsGrouping` is used in two places in formatter_unicode.c and does both places need this change ?

Comment (vstinner) [reply Objects/unicodeobject.c:9384]: Nevermind, I changed my mind and "min_width = Py_MAX(0, min_width);" in _PyUnicode_InsertThousandsGrouping() is just fine.

File: Misc/NEWS.d/next/Core and Builtins/2018-12-22-22-19-51.bpo-35560.9vMWSP.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2018-12-22-22-19-51.bpo-35560.9vMWSP.rst]: You should mention that the only the "n" format is impacted by the bug. I propose:

Fix an assertion error in :func:`format` in debug build for floating point formatting with "n" format, zero padding and small width. Release build is not impacted. Patch by Karthikeyan Singaravelan.

**Review by serhiy-storchaka — APPROVED** [review]

File: Misc/NEWS.d/next/Core and Builtins/2018-12-22-22-19-51.bpo-35560.9vMWSP.rst
Comment (tirkarthi) [reply Misc/NEWS.d/next/Core and Builtins/2018-12-22-22-19-51.bpo-35560.9vMWSP.rst]: Thanks Victor for the suggestion I have reworded the entry.

**Comment by vstinner** [comment]
> vstinner removed the needs backport to 3.6 label a minute ago

Python 3.6 no longer accepts bugfixes: https://devguide.python.org/#status-of-python-branches

**Comment by tirkarthi** [comment]
Thanks @vstinner and @serhiy-storchaka for the review.

**Comment by charris** [comment]
Can this be backported to 3.6? It is causing problems on travisCI with pytest. Note that 3.6 is the default python3 on Ubuntu bionic.

**Comment by tirkarthi** [comment]
I guess this was a problem only with debug builds that has assert statements and as per discussion release builds are not affected. Can you please add a log of travis build that causes this issue in 3.6?

**Comment by charris** [comment]
See https://travis-ci.org/numpy/numpy/jobs/573310511. NumPy is running python3.6-dbg builds on travisCI bionic as one of our CI tests.

**Comment by charris** [comment]
Note that this can affect all users running pytest with the `--durations` flag on python3.6-dbg, it isn't NumPy specific.

**Comment by tirkarthi** [comment]
Thanks for the details. This fix was made during 3.6.8 RC1 which was the last bugfix for 3.6. It is in security fixes mode so I will defer to core devs if we need to backport this.

**Comment by vstinner** [comment]
> Can this be backported to 3.6? It is causing problems on travisCI with pytest. Note that 3.6 is the default python3 on Ubuntu bionic.

Python 3.6 no longer accept bugfixes anymore: https://devguide.python.org/#status-of-python-branches

@ned-deily is the Python 3.6 release manager. Ned: would you be ok to backport this bugfix which cause practical issues on Travis CI?

@charris: Even if Python 3.6 is fixed, I'm not sure that Travis CI is going to get the fix yet. It may take a few weeks or a few months until Travis CI will get the fix. Maybe you should modify pytest tests to skip the test on Python 3.6?

**Comment by charris** [comment]
Thanks for considering a backport.

The workaround for us at this point is to not use `--durations`. It is a nice thing to have, but not critical, and two years from now the testing might move on to 3.7 :) Another possibility is for pytest to rewrite the format expression to not cause the error. I don't why it causes a problem, so any help in that regard would be welcome. See https://github.com/pytest-dev/pytest/issues/5763, which also shows the troublesome line of code.

**Comment by ned-deily** [comment]
As Victor notes, this issue falls outside the criteria for a security-fix branch like 3.6 and, even if made, a backport to 3.6 is unlikely to make a practical difference to your CI testing any time soon. I think it would be better to concentrate on avoiding the problem in your project’s code since that’s something you control.

**Pull request merged**
